In [29]:
import numpy as np
import pandas as pd
from scipy.spatial.distance import cityblock, euclidean, cosine
from sklearn.metrics.pairwise import cosine_similarity
import matplotlib.pyplot as plt
from scipy.spatial.distance import pdist, squareform

# **1. Minkowski Distance (Manhattan and Euclidean)**

In [30]:
# Create a small dataset for similarity calculations
data1 = pd.DataFrame({
    'Movie.1': [1, 5],
    'Movie.2': [2, 3]
}, index=['Alice', 'Bernard'])

data1

,Movie.1,Movie.2
Alice,1,2
Bernard,5,3


## `pdist` - Pairwise Distance in SciPy

`pdist` stands for **pairwise distance** and is a function from SciPy's `spatial.distance` module. It computes the **distance between every pair of rows** in a dataset.

### Parameters:
- **`data`**: A matrix where each row is a point (observation).
- **`metric`**: The type of distance to compute (e.g., `'euclidean'`, `'cityblock'`, `'cosine'`, etc.).


In [31]:
# Calculate Manhattan Distance (L1 norm)
pdist(data1, metric='cityblock')

array([5.])

We can also produce a matrix, this becomes useful when we have more than 2 users

In [32]:
# square matrix
manhattan_dist_matrix_1 = pd.DataFrame(squareform(pdist(data1, metric='cityblock')),
                                     index=data1.index,
                                     columns=data1.index)
manhattan_dist_matrix_1

,Alice,Bernard
Alice,0.0,5.0
Bernard,5.0,0.0


In [33]:
# Calculate Euclidean Distance (L2 norm)
pdist(data1, metric='euclidean')

array([4.12310563])

In [34]:
# Compute Euclidean Distance (L2 Norm)
euclidean_dist_matrix_1 = pd.DataFrame(squareform(pdist(data1, metric='euclidean')),
                                     index=data1.index,
                                     columns=data1.index)
euclidean_dist_matrix_1

,Alice,Bernard
Alice,0.000000,4.123106
Bernard,4.123106,0.000000


# **2. Correlation Calculation**

In [35]:
data1.T.corr(method='pearson')

,Alice,Bernard
Alice,1.0,-1.0
Bernard,-1.0,1.0


The .corr() function in pandas computes the correlation matrix between numerical columns in a DataFrame. It returns a square matrix where each cell represents the correlation between two columns.

### Diagonal Values (1.0):
- Alice's and Bernard's ratings are perfectly correlated with themselves (by definition).

### Off-Diagonal Values (-1.0):
- Indicates a perfect negative correlation between Alice and Bernard, meaning: Alice likes movie 2 better while Bernard likes movie 1 better



# **3. Cosine Distance**

In [36]:
# can also use pdist, note this is distance
cos_dist_matrix_1 = pd.DataFrame(squareform(pdist(data1, metric='cosine')),
                                     index=data1.index,
                                     columns=data1.index)
cos_dist_matrix_1

,Alice,Bernard
Alice,0.000000,0.156339
Bernard,0.156339,0.000000


In [37]:
# can also use cosine_similarity, this is calculating cosine_similarity, which is 1-cosine_distance
cosine_similarity(data1)

array([[1.        , 0.84366149],
       [0.84366149, 1.        ]])

# **4. Handling Missing Data**

In [38]:
data2 = pd.DataFrame({
    'Movie.1': [np.nan, 1],
    'Movie.2': [np.nan, 2],
    'Movie.3': [4, 2],
    'Movie.4': [4, 2],
    'Movie.5': [5, 3]
}, index=['Alice', 'Bernard'])

print("Dataset 2 with missing values:")
print(data2)

Dataset 2 with missing values:
         Movie.1  Movie.2  Movie.3  Movie.4  Movie.5
Alice        NaN      NaN        4        4        5
Bernard      1.0      2.0        2        2        3


## Pearson Distane

The default option is to ignore missing values when calculating correlation.

In [39]:
# Compute pearson distance
pearson_dist2 = data2.T.corr(method='pearson')
pearson_dist2

,Alice,Bernard
Alice,1.0,1.0
Bernard,1.0,1.0


## Cosine distance

In [40]:
# cosine distance: ignoring missing values, only compare the ratings when both users rate
# Remove columns with any missing values
data2_cleaned = data2.dropna(axis=1, how='any')

In [41]:
data2_cleaned

,Movie.3,Movie.4,Movie.5
Alice,4,4,5
Bernard,2,2,3


In [45]:
cos_dist_matrix_2 = pd.DataFrame(squareform(pdist(data2_cleaned, metric='cosine')),
                                     index=data2_cleaned.index,
                                     columns=data2_cleaned.index)
cos_dist_matrix_2

,Alice,Bernard
Alice,0.000000,0.004137
Bernard,0.004137,0.000000


#Compare distance functions after scaling one set of values

Cindy's scores are Alice's plus a constant

In [47]:
data3 = pd.DataFrame({
    'Movie.1': [2, 1, 7],
    'Movie.2': [2, 2, 7],
    'Movie.3': [4, 2, 9],
    'Movie.4': [4, 2, 9],
    'Movie.5': [5, 3, 10]
}, index=['Alice', 'Bernard', 'Cindy'])

print("Dataset with scaled values:")
print(data3)

Dataset with scaled values:
         Movie.1  Movie.2  Movie.3  Movie.4  Movie.5
Alice          2        2        4        4        5
Bernard        1        2        2        2        3
Cindy          7        7        9        9       10


In [48]:

manhattan_dist_matrix_3 = pd.DataFrame(squareform(pdist(data3, metric="cityblock")),
                                     index=data3.index,
                                     columns=data3.index)

manhattan_dist_matrix_3


,Alice,Bernard,Cindy
Alice,0.0,7.0,25.0
Bernard,7.0,0.0,32.0
Cindy,25.0,32.0,0.0


Manhattan distance indicates the distance between Alice and Cindy is 25.

In [50]:
#if you dont use squareform it becomes less intuitive
pdist(data3, metric="cityblock")

array([ 7., 25., 32.])

In [51]:
# Compute Euclidean Distance (L2 Norm)
euclidean_dist_matrix_3 = pd.DataFrame(squareform(pdist(data3, metric="euclidean")),
                                     index=data3.index,
                                     columns=data3.index)
euclidean_dist_matrix_3

,Alice,Bernard,Cindy
Alice,0.000000,3.605551,11.180340
Bernard,3.605551,0.000000,14.422205
Cindy,11.180340,14.422205,0.000000


Euclidean distance indicates the distance between Alice and Cindy is 11.18

In [52]:
# Correlation with scaled values
correlation_matrix_scaled = data3.T.corr(method='pearson')
correlation_matrix_scaled

,Alice,Bernard,Cindy
Alice,1.000000,0.790569,1.000000
Bernard,0.790569,1.000000,0.790569
Cindy,1.000000,0.790569,1.000000


Pearson correlation correctly identifies Cindy as being perfectly correlated with Alice.

In [53]:
#cosine distance
cos_dist_matrix_3 = pd.DataFrame(squareform(pdist(data3, metric='cosine')),
                                     index=data3.index,
                                     columns=data3.index)
cos_dist_matrix_3

,Alice,Bernard,Cindy
Alice,0.000000,0.021561,0.019419
Bernard,0.021561,0.000000,0.022410
Cindy,0.019419,0.022410,0.000000
